ext/mbstring/libmbfl/mbfl/mbfilter.c

```diff
@@ -83,2575 +83,1531 @@
+		/* If we're searching in the reverse direction (from string end going backwards),
+		 * and the offset was negative, then we will search the area _BEFORE_ the offset
+		 * position (with a positive offset, we would search the area after the offset position)
+		 *
+		 * But if the needle is longer than 1 character, and it appears right at the end of
+		 * the area to be searched, it may pass over the offset position, overlapping it
+		 * Expand the area to be searched just enough so that we still find the needle in
+		 * such cases */
+		offset_pointer = offset_pointer + char_offset_to_byte_offset(haystack->encoding,
+			offset_pointer, haystack->val + haystack->len - offset_pointer, needle_len);
+		found_pos = zend_memnrstr(
+			(const char*)haystack->val,
+			(const char*)needle->val, needle->len,
+			(const char*)offset_pointer);
 	}
 
-	/* cleanup */
-	/* dtors should be called in reverse order */
-	i = num;
-	while (--i >= 0) {
-		mbfl_identify_filter_cleanup(&flist[i]);
+	if (found_pos) {
+		return byte_offset_to_char_offset(haystack->encoding, haystack->val, haystack->len,
+			((unsigned char*)found_pos) - haystack->val);
 	}
 
-	efree((void *)flist);
-
-	return encoding;
+	return MBFL_ERROR_NOT_FOUND;
 }
 
 /*
- *  strlen
+ *  substr_count
  */
-static int
-filter_count_output(int c, void *data)
+size_t mbfl_substr_count(mbfl_string *haystack, mbfl_string *needle)
 {
-	(*(size_t *)data)++;
-	return c;
-}
+	ZEND_ASSERT(haystack->encoding == needle->encoding);
 
-size_t
-mbfl_strlen(const mbfl_string *string)
-{
-	size_t len, n, k;
-	unsigned char *p;
-	const mbfl_encoding *encoding = string->encoding;
+	if (mbfl_strlen(needle) == 0) {
+		return MBFL_ERROR_EMPTY;
+	}
 
-	len = 0;
-	if (encoding->flag & MBFL_ENCTYPE_SBCS) {
-		len = string->len;
-	} else if (encoding->flag & (MBFL_ENCTYPE_WCS2BE | MBFL_ENCTYPE_WCS2LE)) {
-		len = string->len/2;
-	} else if (encoding->flag & (MBFL_ENCTYPE_WCS4BE | MBFL_ENCTYPE_WCS4LE)) {
-		len = string->len/4;
-	} else if (encoding->mblen_table != NULL) {
-		const unsigned char *mbtab = encoding->mblen_table;
-		n = 0;
-		p = string->val;
-		k = string->len;
-		/* count */
-		if (p != NULL) {
-			while (n < k) {
-				unsigned m = mbtab[*p];
-				n += m;
-				p += m;
-				len++;
-			}
-		}
-	} else {
-		/* wchar filter */
-		mbfl_convert_filter *filter = mbfl_convert_filter_new(
-		  string->encoding,
-		  &mbfl_encoding_wchar,
-		  filter_count_output, 0, &len);
-		if (filter == NULL) {
-			return (size_t) -1;
-		}
-		/* count */
-		n = string->len;
-		p = string->val;
-		if (p != NULL) {
-			while (n > 0) {
-				(*filter->filter_function)(*p++, filter);
-				n--;
-			}
+	size_t result = 0;
+	unsigned char *search_pos = haystack->val, *found_pos, *e = search_pos + haystack -> len;
+
+	while (search_pos < e) {
+		found_pos = (unsigned char*)zend_memnstr(
+			(const char*)search_pos,
+			(const char*)needle->val, needle->len,
+			(const char*)haystack->val + haystack->len);
```

Comment:
Similarly here, we need to convert to a self-synchronizing encoding to make this work.

Reply:
Yessir.